jsengine-ios/jsengine/WebRequest.swift

```diff
@@ -0,0 +1,121 @@
+            if let tabActive = self?.isTabActive(tabId) {
+                return tabActive
+            }
+            return false
+        }
+        webRequest.setObject(unsafeBitCast(nativeIsWindowActive, AnyObject.self), forKeyedSubscript: "_nativeIsWindowActive")
+        
+        context.setObject(webRequest, forKeyedSubscript: "webRequest")
+    }
+    
+    func shouldBlockRequest(request: NSURLRequest) -> Bool {
+        guard BlockedRequestsCache.sharedInstance.hasRequest(request) == false else {
+            return true
+        }
+        
+        let requestInfo = getRequestInfo(request)
+        if let blockResponse = getBlockResponseForRequest(requestInfo) where blockResponse.count > 0 {
+            BlockedRequestsCache.sharedInstance.addBlockedRequest(request)
+            return true
+        }
+        
+        return false
+    }
+    
+    func newTabCreated(tabId: Int, webView: UIView) {
+        tabs.setObject(webView, forKey: NSNumber(integer: tabId))
+    }
+    
+    //MARK: - Private Methods
+    private func isTabActive(tabId: Int) -> Bool {
+        return tabs.objectForKey(tabId) != nil
+    }
+    
+    private func getBlockResponseForRequest(requestInfo: [String: AnyObject]) -> [NSObject : AnyObject]? {
+        
+        if let requestInfoJsonString = toJSONString(requestInfo) {
+            let onBeforeRequestCall = "System.get('platform/webrequest').default.onBeforeRequest._trigger(\(requestInfoJsonString));"
+            let blockResponse = self.jsContext?.evaluateScript(onBeforeRequestCall).toDictionary()
+            return blockResponse
+        }
+        return nil
+    }
+    
+    private func getRequestInfo(request: NSURLRequest) -> [String: AnyObject] {
+        let url = request.URL?.absoluteString
+        let userAgent = request.allHTTPHeaderFields?["User-Agent"]
+        
+        let isMainDocument = request.URL == request.mainDocumentURL
+        let tabId = getTabId(userAgent)
+        let isPrivate = false
+        let originUrl = request.mainDocumentURL?.absoluteString
+        
+        var requestInfo = [String: AnyObject]()
+        requestInfo["url"] = url
+        requestInfo["method"] = request.HTTPMethod
+        requestInfo["tabId"] = tabId ?? -1
+        requestInfo["parentFrameId"] = -1
+        // TODO: frameId how to calculate
+        requestInfo["frameId"] = tabId ?? -1
+        requestInfo["isPrivate"] = isPrivate
+        requestInfo["originUrl"] = originUrl
+        let contentPolicyType = ContentPolicyDetector.sharedInstance.getContentPolicy(request, isMainDocument: isMainDocument)
+        requestInfo["type"] = contentPolicyType;
+        
+        requestInfo["requestHeaders"] = request.allHTTPHeaderFields
+        return requestInfo
+    }
+    
+    private func getTabId(userAgent: String?) -> Int? {
+        guard let userAgent = userAgent else { return nil }
+        guard let loc = userAgent.rangeOfString("_id/") else {
```

Comment:
In Cliqz Webview, they concatenate the tabId to the userAgent for all requests. Since, at the interceptor level, there is no way to detect which tab issued this request.